README.txt

Handfull of misc olly and ida plugins I have created

IDA_Jscript:
   Script IDA in Javascript. uses IDASrvr to listen to WM_COPYDATA messages.
   script interface is in external process for ease of development. Uses scintinella
   control for syntax highlighting, and intellisense features. See screen shot.   

IDA_JScript_w_DukDbg:
   as above but now using the dukdbg.ocx as full on javascript debugger. The ocx 
   is usable, but binary compatability has not been set yet so you may have to recompile
   one or both projects to use yourself.
   https://github.com/dzzie/duk4vb/tree/master/vb_examples/full_ocx

olly_hittrace: 
   You set breakpoints in the UI and it will then run
   the app automating it and logging which ones were hit.

Olly_module_bpx:
   allow you to set breakpoints within modules which have not yet been loaded.

Olly_vbscript:
   vbscript automation capability for olly including working across breakpoint events.
   also includes ollysync which is udp based remote control to query and control olly
   also can be used to send display sync commands to idavbscript while stepping in olly

OllySrvr:
   wm_copydata IPC server running in olly. allows you to query data from current process
   and sync display. 

IDASrvr: 
   wm_copydata IPC server running in IDA. allows you to send commands to IDA from another
   process to query data and control interface display. used from my uGrapher wingraph32 
   replacement so when you click on a node it will navigate to the function in IDA.
   WM_COPYDATA was chosen because it is synchronous. Sample projects in many languages
   and also contains an ActiveX client library which is very easy to use.

IDASRVR2:
   in progress...all offsets in this build will support x64 for both plw and p64 builds
   this is on the path for adding in p64 support for both 32bit and then 64 bit IDA versions
   grumble grumble.


uGrapher:
   replaces default wingraph32, uses uDrawGraph. graphing up, still needs features. once
   idasrvr working again, it can navigate to nodes as you click then in the graph..not 
   sure i like uDraw graph after having used some other activeX controls. I already have
   this working and more in another way using the socket based ida control in idavbscript.
   but the uDrawGraph classes are nice and debugged. ported from pedrams python work on 
   openrce.

IdaVbScript:
   ton of small tools for IDA all thrown into one interface. see screenshot.png for 
   details. also include a udp based command socket to send commands and receive data 
   from IDA from external apps. Used by olly_vbscript and wingraph32 replacement. 
   (not uGrapher)

IdaUdpBridge:
   this replaces the udp command socket in idavbscript which was crashy. This will listen
   on udp 3333 for command messages and then proxy them to IDA. It uses the IDASrvr plugin
   and the \IDASrvr\ActiveX_clientLib. Used by olly_vbscript (ollysync) and wingraph32 
   replacement. 

Wingraph32:
   another wingraph32 replacement using a 3rd party graphing ocx (see readme), see screen 
   shot for features. can also navigate IDA to selected nodes using the UDP socket in 
   idavbscript plugin.

gleegraph:
   c# wingraph32 replacement using Microsoft GLEE library (free for non-commercial use)
   click on a node to navigate IDA to it. Right click menu to also rename functions within
   IDA, or prefix all nodes below. Uses IDASrvr see screen shot.

wininet_hooks:
   logs http/https traffic with api injection to capture raw data regardless of ssl. 

IDASrvr2/changes.txt

activex client lib switch to spsubclass so each class can hook its own window?
or support multiple hooks...

instead of memfile we could use WM_GETTEXT, WM_SETTEXT and pass a texbox hwnd in as arg1 since it is 32bit safe

QuickCalls changed: 
   18/28 quick calls depends on arg1 being an address..not cool for x64 addresses w/ 32bit caller
   6 of these do not exist in traditional call mechanism *

 these three can be used traditionally still from 32bit code, x64 must use memfile...
   case 8: // imgbase changed now memfile output
   case 14: //funcstart:funcIndex changed to memfile output
   case 15: //funcend:funcIndex - changed to memfile output

 wont work for quickcall at all without mods
   case 1: // jmp:lngAdr
   case 10: //readbyte:lngva - 
   case 11: //orgbyte:lngva  - 
   case 20: //undefine:offset - 
   case 22: //hide:offset - 
   case 23: //show:offset - 
   case 24: //remname:offset - 
   case 25: //makecode:offset   - 
   case 32: //funcindex:va 
   case 33: //nextea:va  should this return null if it crosses function boundaries? yes probably... 
   case 34: //prevea:va  should this return null if it crosses function boundaries? yes probably...  
   case 37: //screenea: 

 not implemented in regular call yet...
  case 44: return isCode(getFlags(tmp)); - 
  case 45:return isData(getFlags(tmp)); - 
  case 46:return decode_insn(tmp);- 
  case 47:return get_long(tmp);- 
  case 48:return get_word(tmp);- 


QuickCalls removed..


all addresses received by HandleMsg now use _atoi64

I should be able to compile this as a p64 now...and maybe as a 64bit version of the plugin for ida 7 64bit..
still no client updates yet...

clients with an underscore have not yet been worked on


regular calls changed:

	q  14 funcstart:funcIndex[:hwnd]  - x64 requires hwnd, legacy 32bit code still ok
	q  15 funcend:funcIndex[:hwnd]  - x64 requires hwnd, legacy 32bit code still ok
	q  33 nextea:va[:hwnd]   - x64 requires hwnd, legacy 32bit code still ok
	q  34 prevea:va[:hwnd]  - x64 requires hwnd, legacy 32bit code still ok
        q* 37 screenea:[:hwnd]  - x64 requires hwnd, legacy 32bit code still ok


IDASrvr2/c_client/c_client.cpp

#include <stdio.h>
#include <conio.h>
#include <string.h>
#include <windows.h>
#include <stdlib.h>

const int WM_DISPLAY_TEXT = 3;

//this is a super quick and dirty demo of a C client..

typedef struct{
    int dwFlag;
    int cbSize;
    int lpData;
} cpyData;


HWND hServer;
WNDPROC oldProc;
cpyData CopyData;

int  m_debug = 0;
int  m_ServerHwnd = 0;
char* IDA_SERVER_NAME = "IDA_SERVER2";
char m_msg[2020];
bool received_response = false;
UINT IDA_QUICKCALL_MESSAGE;
int IDA_HWND=0;

//quick call offers about 3x performance boost over original..
enum quickCallMessages{
	qcmJmpAddr = 1, // jmp:lngAdr
	qcmJmpRVA = 7, // jmp_rva:lng_rva
	qcmImgBase = 8, // imgbase 
	qcmReadByte = 10, //readbyte,lngva 
	qcmOrgByte = 11, //orgbyte,lngva 
	qcmRefresh = 12, // refresh,
	qcmNumFuncs = 13, // numfuncs 
	qcmFuncStart = 14, //funcstart,funcIndex 
	qcmFuncEnd = 15, //funcend,funcIndex 
	qcmUndef = 20, //undefine,offset
	qcmHide = 22, //hide,offset
	qcmShow = 23, //show,offset
	qcmRemName = 24, //remname,offset
	qcmMakeCode = 25, //makecode,offset
	qcmFuncIdx = 32, //funcindex,va 
	qcmNextEa = 33, //nextea,va   
	qcmPrevEa = 34 //prevea:va   
};
		

#pragma warning (disable:4996)

HWND ReadReg(char* name){ //todo support multiple IDA idb's and not just last one opened...

	 char* baseKey = "Software\\VB and VBA Program Settings\\IPC\\Handles";
	 char tmp[20] = {0};
     unsigned long l = sizeof(tmp);
	 HKEY h;
	 
	 RegOpenKeyExA(HKEY_CURRENT_USER, baseKey, 0, KEY_READ, &h);
	 RegQueryValueExA(h, name, 0,0, (unsigned char*)tmp, &l);
	 RegCloseKey(h);

	 return (HWND)atoi(tmp);
}

int SendTextMessage(int hwnd, char *Buffer, int blen) 
{
		  char* nullString = "NULL";
		  if(blen==0){ //in case they are waiting on a message with data len..
				Buffer = nullString;
				blen=4;
		  }
		  if(m_debug){
			  printf("Trying to send message to %x size:%d\n", hwnd, blen);
			  printf(" msg> %s\n", Buffer);
		  }
		  if(Buffer[blen] != 0) Buffer[blen]=0; ;
		  cpyData cpStructData;  
		  cpStructData.cbSize = blen;
		  cpStructData.lpData = (int)Buffer;
		  cpStructData.dwFlag = 3;
		  return SendMessage((HWND)hwnd, WM_COPYDATA, (WPARAM)hwnd,(LPARAM)&cpStructData);  
}  

int SendIntMessage(int hwnd, __int64 resp){
	char tmp[30]={0};
	sprintf(tmp, "%llu", resp);
	if(m_debug) printf("SendIntMsg(%d, %s)", hwnd, tmp);
	return SendTextMessage(hwnd,tmp, sizeof(tmp));
} 

void HandleMsg(char* m){
	//Message Received from IDA do stuff here...
	printf("%s\n", m);
}

/* old method */		//these next 2 are a very simple implementation..SendMessage automatically blocks so this works...
__int64 ReceiveInt(char* command, int hwnd){
	memset(m_msg,0,2020);
	received_response = false;
	SendTextMessage(hwnd,command,strlen(command)+1);
	return _atoi64(m_msg);
}

int NewReceiveInt(char* command, int hwnd){
	return SendTextMessage(hwnd,command,strlen(command)+1);
}

char* ReceiveText(char* command, int hwnd){
	memset(m_msg,0,2020);
	received_response = false;
	SendTextMessage(hwnd, command,strlen(command)+1);
	return m_msg;
}


LRESULT CALLBACK WindowProc(HWND hwnd,UINT uMsg,WPARAM wParam,LPARAM lParam){

		if( uMsg != WM_COPYDATA) return 0;
		if( lParam == 0) return 0;
		
		memcpy((void*)&CopyData, (void*)lParam, sizeof(cpyData));
    
		if( CopyData.dwFlag == 3 ){
			if( CopyData.cbSize >= sizeof(m_msg) ) CopyData.cbSize = sizeof(m_msg)-1;
			memcpy((void*)&m_msg[0], (void*)CopyData.lpData, CopyData.cbSize);
			if(m_debug)	printf("Message Received: %s \n", m_msg); 
			received_response = true;
		}
			
    return 0;
}

//some quick call messages now require reading memfile which we havent setup in this example...
//might eliminate those from the quick call..
int QuickCall(quickCallMessages msg, int arg1 =0 ){
	return SendMessage( (HWND)IDA_HWND, IDA_QUICKCALL_MESSAGE, msg, arg1);
}

int main(int argc, char* argv[])
{
 
	system("cls");

	m_ServerHwnd = (int)CreateWindowA("EDIT","MESSAGE_WINDOW", 0, 0, 0, 0, 0, 0, 0, 0, 0);
	oldProc = (WNDPROC)SetWindowLongA((HWND)m_ServerHwnd, GWL_WNDPROC, (LONG)WindowProc);

	IDA_HWND = (int)ReadReg(IDA_SERVER_NAME);
	if(!IsWindow((HWND)IDA_HWND)) IDA_HWND = 0;

	IDA_QUICKCALL_MESSAGE = RegisterWindowMessageA("IDA_QUICKCALL2");

	if( m_ServerHwnd == 0){
		printf("Could not create listener window to receive data on exiting...\n");
		printf("Press any key to exit..");
		getch();
		return 0;
	}

	if( IDA_HWND==0 ){
		printf("IDA Server window not found exiting...\n");
		printf("Press any key to exit..");
		getch();
		return 0;
	}
	
	printf("Listening for responses on hwnd: %d\n", m_ServerHwnd); 
	printf("Active IDA hwnd: %d\n", IDA_HWND);

	char buf[255];
	__int64 ret = 0;
    char* sret = 0;
	LARGE_INTEGER start , end, freq;
	
	QueryPerformanceFrequency(&freq);
	printf("1 tick = 1/%d seconds\n\n", freq.QuadPart);

	sprintf(buf,"loadedfile:%d", m_ServerHwnd);
	sret = ReceiveText(buf,IDA_HWND); 	
    printf("Loaded IDB: %s\n", sret);

	sprintf(buf,"numfuncs:%d", m_ServerHwnd);
	
	QueryPerformanceCounter(&start);
	ret = ReceiveInt(buf, IDA_HWND); //two copydata's  + sprintf + string to int
	QueryPerformanceCounter(&end);
	
	printf("Function Count: %d  (org method %d ticks)\n", ret, end.QuadPart-start.QuadPart);

	QueryPerformanceCounter(&start);
	ret = NewReceiveInt("numfuncs", IDA_HWND);
	QueryPerformanceCounter(&end);
	
	printf("Function Count: %d  (new method %d ticks)\n", ret, end.QuadPart-start.QuadPart);

	QueryPerformanceCounter(&start);
	ret = QuickCall(qcmNumFuncs);
	QueryPerformanceCounter(&end);
	
	printf("Function Count: %d  (quick call %d ticks)\n", ret, end.QuadPart-start.QuadPart);
	
	sprintf(buf,"funcstart:0:%d", m_ServerHwnd);
	int funcStart = ReceiveInt(buf, IDA_HWND);
	printf("\nFirst Func Start: 0x%x\n", funcStart);

	sprintf(buf,"funcend:0:%d", m_ServerHwnd);
	ret = ReceiveInt(buf, IDA_HWND);
	printf("First Func End: 0x%x\n", ret);

	sprintf(buf,"getasm:%d:%d", funcStart, m_ServerHwnd);
	sret = ReceiveText(buf,IDA_HWND); 	
    printf("First Func Disasm[0]: %s\n", sret);
	
	printf("Press any key to exit..");
	getch();
	
	return 0;

}